eclair-core/src/main/scala/fr/acinq/eclair/channel/Commitments.scala

```diff
@@ -98,16 +98,17 @@ case class Commitments(channelVersion: ChannelVersion,
     if (localParams.isFunder) {
       // The funder always pays the on-chain fees, so we must subtract that from the amount we can send.
       val commitFees = commitTxFeeMsat(remoteParams.dustLimit, reduced)
-      // the funder needs to keep an extra reserve to be able to handle fee increase without getting the channel stuck
-      // (see https://github.com/lightningnetwork/lightning-rfc/issues/728)
-      val funderFeeReserve = htlcOutputFee(2 * reduced.feeratePerKw)
-      val htlcFees = htlcOutputFee(reduced.feeratePerKw)
-      if (balanceNoFees - commitFees < offeredHtlcTrimThreshold(remoteParams.dustLimit, reduced)) {
+      // the funder needs to keep an extra buffer to be able to handle a x2 feerate increase and an additional htlc to avoid
+      // getting the channel stuck (see https://github.com/lightningnetwork/lightning-rfc/issues/728).
+      val funderFeeBuffer = commitTxFeeMsat(remoteParams.dustLimit, reduced.copy(feeratePerKw = 2 * reduced.feeratePerKw)) + htlcOutputFee(2 * reduced.feeratePerKw)
+      if (balanceNoFees - commitFees.max(funderFeeBuffer) < offeredHtlcTrimThreshold(remoteParams.dustLimit, reduced)) {
         // htlc will be trimmed
-        (balanceNoFees - commitFees - funderFeeReserve).max(0 msat)
+        (balanceNoFees - commitFees.max(funderFeeBuffer)).max(0 msat)
       } else {
         // htlc will have an output in the commitment tx, so there will be additional fees.
-        (balanceNoFees - commitFees - funderFeeReserve - htlcFees).max(0 msat)
+        val commitFees1 = commitFees + htlcOutputFee(reduced.feeratePerKw)
+        val funderFeeBuffer1 = funderFeeBuffer + htlcOutputFee(2 * reduced.feeratePerKw)
```

Comment:
You are adding an `htlcOutputFee` to `funderBufferFee` which already had an additional `htlcOutputFee`?

Reply:
This is damn confusing. An additional comment may help (I'm adding it).

In `funderFeeBuffer` we take into account the requirement to handle a *future* additional HTLC at `2 * feeratePerKw`.
In this `else` branch, we are adding a *present* HTLC into the commitment, so we need to take its fee into account (just as we add it to `commitFees` to obtain `commitFees1`). And since the fee buffer uses `2 * feeratePerKw`, this is why we use that value inside `htlcOutputFee`.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1364/commits/fbe4529c97a9c2b6c1056f49fe76551009708b94

Reply:
I still don't get it. If we are in the *trimmed* case, then why do we include an `htlcOutputFee` in `funderFeeBuffer`?

Reply:
In this `else` branch we are in the *untrimmed* case, so we take the current htlc into account.
In `funderFeeBuffer` we kept some room for a *future, potential* untrimmed HTLC at twice the current feerate (this is really just keeping a buffer for an unknown future).
Is it a bit more clear?

Reply:
Ok I think I'm understanding your comment a bit better.

In all cases (both trimmed and untrimmed), we need to reserve the cost of adding a future untrimmed HTLC at twice the feerate (which we do when we calculate `funderFeeBuffer`).

If after that, we still have enough balance to send untrimmed HTLCs (which is what happens in this `else` branch), we need to take into account the cost of that untrimmed HTLC. This is why we add the htlc fee to `commitFees` to get `commitFees1`. Similarly, we need to add this HTLC to the funder fee buffer, but at twice the feerate, which is how we calculate `funderFeeBuffer1`. It may look like we already accounted for it when calculating `funderFeeBuffer`, but this is not the same HTLC! In `funderFeeBuffer` we added the cost for an HTLC in the future, whereas in `funderFeeBuffer1` we need to add the cost for an HTLC now (`availableBalanceForSend` considers that you're going to try to send that HTLC now).

Reply:
I get it now. Thanks!